fallback_builtins.h

```diff
@@ -4,41 +4,43 @@
 #if defined(_MSC_VER) && !defined(__clang__)
 #if defined(_M_IX86) || defined(_M_AMD64) || defined(_M_IA64) ||  defined(_M_ARM) || defined(_M_ARM64)
 
+#include <assert.h>
 #include <intrin.h>
 #ifdef X86_FEATURES
 #  include "arch/x86/x86_features.h"
 #endif
 
-/* This is not a general purpose replacement for __builtin_ctz. The function expects that value is != 0
- * Because of that assumption trailing_zero is not initialized and the return value of _BitScanForward is not checked
+/* This is not a general purpose replacement for __builtin_ctz. The function expects that value is != 0.
+ * Because of that assumption trailing_zero is not initialized and the return value is not checked.
+ * Tzcnt and bsf give identical results except when input value is 0, therefore this can not be allowed.
+ * If tzcnt instruction is not supported, the cpu will itself execute bsf instead.
+ * Performance tzcnt/bsf is identical on Intel cpu, tzcnt is faster than bsf on AMD cpu.
  */
 static __forceinline int __builtin_ctz(unsigned int value) {
-#ifdef X86_FEATURES
-#  ifndef X86_NOCHECK_TZCNT
-    if (x86_cpu_has_tzcnt)
-#  endif
-        return (int)_tzcnt_u32(value);
-#endif
+    assert(value != 0);
+# if (_MSC_VER < 1700)
     unsigned long trailing_zero;
     _BitScanForward(&trailing_zero, value);
     return (int)trailing_zero;
+# else
+    return (int)_tzcnt_u32(value);
+# endif
```

Comment:
``tzcnt`` (``_tzcnt_u32``, ``_tzcnt_u64``) is only available on x86:

```c++
#ifdef X86_FEATURES
    return (int)_tzcnt_u32(value);
# else
    unsigned long trailing_zero;
    _BitScanForward(&trailing_zero, value);
    return (int)trailing_zero;
# endif
```

Reply:
Fixed